### PR TITLE
Added Player chat simulation method. Adds #733

### PR DESCRIPTION
**Pull request by me4502**
[API](https://github.com/SpongePowered/SpongeAPI/pull/1283) | [Common](https://github.com/SpongePowered/SpongeCommon/pull/895)

This PR adds a method to Player, that allows for the simulation of chat.

The exact reasoning of it can be seen in the corresponding issue, https://github.com/SpongePowered/SpongeAPI/issues/733

This method returns the event so that the plugin can see what modifications have been made by other plugins, and to see if it has been cancelled.

Test code,

``` java
    @Listener
    public void onInteract(InteractBlockEvent.Secondary event, @First Player player) {
        player.simulateChat(Text.of("Hello there"));
    }

    @Listener
    public void onMessage(MessageChannelEvent.Chat chat, @First Player player) {
        Sponge.getServer().getBroadcastChannel().send(Text.of("Hello, " + player.getName() + '!'));
    }
```

### CODE REVIEW

**Comment by me4502** [comment]
As a potential discussion point - should the plugin be added as the root cause, rather than the player? I wasn't sure about this as it'd require passing the plugin - as well as possibly confusing plugins that expect the root cause to be a player.

File: src/main/java/org/spongepowered/api/entity/living/player/Player.java [hunk 1]
```diff
@@ -119,6 +120,14 @@
     boolean isChatColorsEnabled();
 
     /**
+     * Simulates a chat message from a player.
+     *
+     * @param message The message to send
+     * @return The event thrown from the message
+     */
+    MessageChannelEvent.Chat chat(Text message);
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: This should accept `String` in my opinion. Clients send a `String` only, and this method should be used to simulate a client sending a `String` to the server.

Comment (me4502) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: That would however limit what the plugin can do. It feels like a severe limitation for no reason other than to prevent them doing something the vanilla client can't.

Comment (kashike) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: It would limit it... how? You convert to a legacy text as it is, which limits things as is.

Comment (me4502) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: That was one part I wish to change, I'd need to discuss it with people first - I wanted to move chat handling over to common, and handle it properly there as a Text.

As well as that, I want to discuss how it'd be handled on a client plugin - which is why I've currently put the code in EntityPlayerMP.

Comment (ryantheleach) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: Maybe the client gains the ability to bold text in the future?

Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: There should be a hint whether:
- This actually sends messages to someone or not
- Whether the event is already processed or not
- Whether you can send commands

Comment (me4502) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: @ST-DDT What do you mean by "This actually sends messages to someone or not"?

File: src/main/java/org/spongepowered/api/entity/living/player/Player.java [hunk 2]
```diff
@@ -119,6 +120,17 @@
...
     /**
+     * Simulates a chat message from a player.
+     *
+     * <p>This method sends a message as if it came from this player.
+     * It does not send a message to this player.</p>
+     *
+     * @param message The message to send
+     * @return The event that was thrown from sending the message
+     */
+    MessageChannelEvent.Chat chat(Text message);
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: Your changes fixed most of my concerns. However it still does not explicitly states whether this method can be used to send commands.
Maybe there should be a hint which method can be used to send commands because both features are somehow related.

Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: > It does not send a message to this player.

I'm not quite sure about this one. Because it implies that the used MessageChannel explicitly excludes the given player. Although the player would normally get a formatted version of what he just typed.

Comment (me4502) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: I'll rephrase it to "To send a message to this player, see sendMessage(Text)"

Comment (kashike) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java]: Perhaps call this `simulateChat` instead?

File: src/main/java/org/spongepowered/api/entity/living/player/Player.java [hunk 3]
```diff
@@ -119,6 +123,24 @@
     boolean isChatColorsEnabled();
 
     /**
+     * Simulates a chat message from a player.
+     *
+     * <p>This method sends a message as if it came from this player.
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/entity/living/player/Player.java:137]:
``` java
     * <p>This method sends a message as if it came from this player.
     * To send a message to this player instead, see
     * {@link MessageReceiver#sendMessage(Text)} or
     * {@link ChatTypeMessageReceiver#sendMessage(ChatType, Text)}.</p>
     *
     * <p>Commands cannot be sent using this method. To send commands, use
     * {@link CommandManager#process(CommandSource, String)}.</p>
     *
     * <p>If text formatting is not supported in the implementation
     * it will be displayed as plain text.</p>
```

**Comment by Minecrell** [comment]
@me4502 IMO the plugin shouldn't be the root/source of the cause of the event, but at least part of it, so other plugins can identify the plugin sending the message. Maybe the methods could also take a `Cause` directly so the plugins could add more context if they wanted to?

**Comment by me4502** [comment]
Okay, I've added a Cause to the method. That makes the most sense IMO, and is inline with how setBlock etc work.
